Account Settings: Primary Blog Picker

jleandroperez opened the pull request:
#### Details:
In this PR we'll enhance Account Settings, so that the user would be able to update his Default Blog, right within the app.

@koke may i bother you with a review?. We'll deal with the Refresh bug in another PR, since that's already (also) affecting the `Web Address` edition.

Thanks in advance!

Needs Review: @koke
Closes #4822

#### Testing: WordPress.com Account
1. Launch WPiOS and log into a WordPress.com account
2. Open the Settings tab
3. Tap over Account Settings > Primary Blog
4. Update your primary blog

Please, verify that the new setting is also reflected in WordPress.com.

koke commented on WordPress/Classes/ViewRelated/Me/AccountSettingsViewController.swift:
```diff
@@ -83,11 +96,44 @@ private struct AccountSettingsController: SettingsController {
+            row in
+
+            let selectorViewController = BlogSelectorViewController(selectedBlogObjectID: primarySite?.objectID,
+                selectedCompletion: { self.updatePrimarySite($0) },
```
You'll want `[unowned self] in` here

koke replied:
And in **a lot** of other places :scream:

koke replied:
It's a `struct`, forget I said anything :speak_no_evil:

jleandroperez replied:
The magic of `struct` :smile:. Nevertheless, thanks for spotting that. Would have been a nasty retain cycle!

koke commented:
About the changes in `AccountSettingsViewController`:

- If you use the siteID (`Int`) instead of the objectID (`NSManagedObjectID`) everything is simpler: no need to have a context. The view controller shouldn't touch Core Data at all.
- `mapViewModel` still needs to get the blog name somehow, so I'd add a helper method to `AccountSettingsService`, for instance `primarySiteNameForSettings(_) -> String`, and have that call `BlogService`. With this approach, the VC only receives a String, instead of dealing with managed objects.

koke commented:
Couple issues:

- When you tap on the already selected blog nothing happens. This is inconsistent with selecting any other blog taking you back to the settings screen
- Layout broken on iPad landscape
![simulator screen shot 18 feb 2016 20 19 00](https://cloud.githubusercontent.com/assets/8739/13155215/fd4d0476-d67c-11e5-978b-7cb258f17a13.png)

jleandroperez commented:
@koke although i *absolutely love* the idea of wrapping the Core Data interaction, i'm afraid that updating `BlogSelectorViewController` (used by the Editor to pick the target site, as well) might break things.

Using *only* **blogID** or **dotComID** will break in the `Multiple self hosted websites, without Jetpack` scenario, since both, `blogID` and `dotComID` [are effectively zero, as seen in this screenshot](https://cloudup.com/cX0l2EgRzK6).

This also means that even if we moved the `Primary Site` code over to `AccountSettingsService`, we'd still need to deal with CoreData within `AccountSettingsController`. Unless, right, we only passed thru the `NSManagedObjectID` and moved the `updatePrimarySite` method over to AccountSettingsService as well.

Thoughts?

Thank you!

koke commented:
> Using only blogID or dotComID will break in the Multiple self hosted websites, without Jetpack scenario,...

This is a good point, but then I realized, should the primary site selector show those sites at all? If you don't have a `dotComID` for the blog, can you even set it as a primary in the API?

jleandroperez commented:
@koke not the Primary Site picker. But we're reusing the Site Picker component in the Post Editor, and i'd really love not to hack that guy to have two behaviors (or worse, duplicate that code!)

jleandroperez commented:
@koke all comments addressed:

- iPad should look good now
- Tapping on the preselected row should dismiss the controller
- Primary Site Name helper added to * AccountSettingsService*
- Nuked CoreData interaction in *AccountSettingsController*.

Thanks for the feedback / brainstorming loop sir!. Ready for another round!

koke commented on WordPress/Classes/Services/AccountSettingsService.swift:
```diff
@@ -151,6 +151,13 @@ class AccountSettingsService {
         return notificationObserver.map(getSettings).startWith(getSettings())
     }
 
+    func primarySiteNameForSettings(settings: AccountSettings?) -> String {
```
Why do you accept an optional AccountSettings here? There's no point in calling this method with a `nil` value

jleandroperez replied:
@koke the reason why that method has an optional is that `mapViewModel`'s [signature uses an optional as well, and this is consumed by that helper](https://github.com/wordpress-mobile/WordPress-iOS/blob/issue/4822-primary-blog-selector/WordPress/Classes/ViewRelated/Me/AccountSettingsViewController.swift#L39).

I'd be happy to kill the optional, but this leads me to a cross question for you: why is `mapViewModel` relying on an optional model?

koke replied:
I'm not sure that's a good reason. I think public methods shouldn't be that coupled to their callers. if settings is nil, there's no point in calling `primarySiteNameForSettings` since the return will always be `nil`. That's what `map` is for. As for the return value, if you return a non-optional string it hides the fact that the method might fail: it will return a successful (empty) value even if it can't find a blog with the given ID. 

> why is mapViewModel relying on an optional model?

Because we might not always have the `settings` for an account. I think it was a bit of a shortcut, and I'm exploring [some alternatives](https://github.com/wordpress-mobile/WordPress-iOS/blob/db02ce6545cb2205f1e62d89f463a9669a77a011/WordPress/Classes/ViewRelated/Plans/PlanListViewController.swift#L73) in plans with an `enum` view model that has Loading/Ready/Error states.

Although in this particular case, the design called for showing the table with empty values while loading or after an error, so the mapping for all the states would look too similar.

Oh, I just remembered what the "shortcut" is: when `settings == nil`, the rows shouldn't be `EditableTextRow` but `TextRow` objects, as you shouldn't be able to edit them until we have some data.

koke commented on WordPress/Classes/Services/AccountSettingsService.swift:
```diff
@@ -151,6 +151,13 @@ class AccountSettingsService {
 
+    func primarySiteNameForSettings(settings: AccountSettings?) -> String {
+        let service = BlogService(managedObjectContext: context)
+        let blog = service.blogByBlogId(settings?.primarySiteID)
```
I'd add a `guard` to check if `primarySiteID` is `nil` and bail early.

jleandroperez replied:
Good call there, updating, thanks!

koke commented on WordPress/Classes/Services/AccountSettingsService.swift:
```diff
@@ -151,6 +151,13 @@ class AccountSettingsService {
+        let service = BlogService(managedObjectContext: context)
+        let blog = service.blogByBlogId(settings?.primarySiteID)
+        
+        return blog?.settings?.name ?? String()
```
Why not return an optional `String?` instead of an empty one. And more of a style preference I guess but I'm curious, why `String()` and not `""`?

jleandroperez replied:
@koke we cannot really return an optional here, because this is consumed by `EditableTextRow`, and that class requires `String`.

Regarding the `String()` over `""`, it's a personal decision: i'd rather being explicit. Literal vs new object instantiation.

koke replied:
> this is consumed by EditableTextRow, and that class requires String

Sure, I just meant moving the `?? String()` to the calling code in `mapViewModel`, like the rest of the optional fields.

koke replied:
> Regarding the String() over "", it's a personal decision: i'd rather being explicit. Literal vs new object instantiation

Definitely a personal preference, but IMO it doesn't get more explicit than `""` to say "empty string". When I see that, it's like using `Int()` to say `0`.

koke commented:
The blog selector shows all blogs in the app, not just those associated with the account. If you have any self hosted, you can select that one, and get an API error:

> Error saving account settings change Error Domain=com.wordpress.api Code=0 "You must supply an ID for one of your own sites for primary_site_ID." UserInfo={WordPressComApiErrorCodeKey=invalid_input, NSLocalizedDescription=You must supply an ID for one of your own sites for primary_site_ID.}

koke commented:
OK, I'm purposely trying to break this but:

- Sign in with a wp.com user, go to Me
- From a browser, create a new site (or invite the account to an existing one). Set this new site as the primary site.
- Back in the app, and without going to the sites list, tap on account settings
- Primary site is blank
- When you tap on it, it crashes :boom: 

On `-[BlogSelectorViewController viewWillAppear]`:

```
    NSManagedObject *obj = [self.resultsController.managedObjectContext objectWithID:self.selectedObjectID]; // selectedObjectID is nil
```

jleandroperez commented:
@koke thanks so much for spotting that crash!

- Only Dotcom + Jetpack sites should now show up in the Primary Site Picker
- The crash is gone now. Blogs will get sync'ed, and the list refreshed, automatically
- All of your code-y comments have been addressed (re: optionals)

Ready for another round sir!
Thank you!!

koke commented:
> Only Dotcom + Jetpack sites should now show up in the Primary Site Picker

It's not about .com/Jetpack only, it's about only blogs for that account. If I have a Jetpack site connected to a second account, it's available as a primary site, but when you select it:

>  Error saving account settings change Error Domain=com.wordpress.api Code=0 "You must supply an ID for one of your own sites for primary_site_ID.

jleandroperez commented:
@koke sorry about that, so many corner cases, thanks for catching that one as well.

Everything should be good now!

koke commented:
I'd comment that you might not need to check `jetpackAccount`, but I think I just need to kill that for once and simplify the whole thing.

I'd rather pass the account ID than setting a `displaysOnlyDefaultAccountSites` in case we want to support multiple accounts in the future (I want), but we can change that when we do it.

Last, just an FYI, in case you didn't know, but I'm fine with the current code. You could use `NSCompoundPredicate` there to and/or predicates here:

```
predicate = [predicate stringByAppendingString:@" AND (account == %@ OR jetpackAccount == %@)"];
```

:shipit:

jleandroperez commented:
Keeping the old property (*just in case*), until we effectively nuke it out.

Thanks a lot @koke for the *epic* review!!!

astralbodies commented:
![giphy](https://cloud.githubusercontent.com/assets/373903/13613514/0a4d9898-e532-11e5-92c4-d16c132bf72f.gif)
